src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs

```diff
@@ -260,6 +282,50 @@ private async Task<Document> InlineTemporaryAsync(Document document, VariableDec
             return updatedDocument;
         }
 
+        private static bool MayContainSideEffects(SyntaxNode expression)
+        {
+            // Checks to see if inlining the temporary variable may change the code's semantics. 
+
+            if (expression == null)
+            {
+                return true;
+            }
+
+            if (expression.IsKind(SyntaxKind.ParenthesizedExpression, out ParenthesizedExpressionSyntax parenthesizedExpression))
+            {
+                expression = parenthesizedExpression.WalkDownParentheses();
+            }
+
+            var descendantNodesAndSelf = expression.DescendantNodesAndSelf();
+
+            // e.g.:
+            //     var [||]c = new C();
+            //     c.P = 1;
+            //     var x = c;
+            // After refactoring:
+            //     new C().P = 1;
+            //     var x = new C();
+            if (descendantNodesAndSelf.Any(n => n.IsKind(SyntaxKind.ObjectCreationExpression)))
+            {
+                return true;
+            }
+
+            // e.g. - let method M return a new instance of an object containing property P:
+            //     var [||]c = M();
+            //     c.P = 0;
+            //     var x = c;
+            // After refactoring:
+            //     M().P = 0;
+            //     var x = M();
+            if (descendantNodesAndSelf.Any(n => n.IsKind(SyntaxKind.InvocationExpression)))
+            {
+                return true;
+            }
```

Comment:
more efficient to just do `descendantNodesAndSelf.Any(n => n.IsKind(SyntaxKind.ObjectCreationExpression) || n.IsKind(SyntaxKind.InvocationExpression))`